File: sdk/js/src/transports/mqtt.ts
/**
 * GRAFT MQTT Transport — Node.js + Browser
 *
 * Uses the `mqtt` npm package. Message-oriented (no COBS needed).
 * Install: npm install mqtt
 *
 * Topic schema:
 *   graft/{deviceId}/cmd/{typeHex}  — host publishes commands
 *   graft/{deviceId}/evt/{typeHex}  — device publishes events
 *   graft/{deviceId}/hello          — retained HELLO_RESP
 *   graft/{deviceId}/status         — online/offline (LWT)
 */

import type { GraftTransport } from './transport.js'

export interface MQTTTransportOptions {
  /** MQTT broker URL (e.g., 'mqtt://broker.local' or 'ws://broker.local:9001') */
  broker: string
  /** Device ID for topic routing */
  deviceId: string
  /** MQTT username (optional) */
  username?: string
  /** MQTT password (optional) */
  password?: string
  /** Default QoS for commands (default: 1) */
  qos?: 0 | 1 | 2
}

export class MQTTTransport implements GraftTransport {
  private _options: MQTTTransportOptions
  private _client: any = null
  private _handler: ((packet: Uint8Array) => void) | null = null
  private _connected = false
  private _prefix: string

  readonly needsCOBS = false

  constructor(options: MQTTTransportOptions) {
    this._options = options
    this._prefix = `graft/${options.deviceId}`
  }

  get connected(): boolean {
    return this._connected
  }

  async connect(): Promise<void> {
    const mqtt = await import('mqtt')

    const connectOpts: any = {}
    if (this._options.username) connectOpts.username = this._options.username
    if (this._options.password) connectOpts.password = this._options.password

    this._client = await mqtt.connectAsync(this._options.broker, connectOpts)
    this._connected = true

    // Subscribe to device event topics
    await this._client.subscribeAsync([
      `${this._prefix}/evt/#`,
      `${this._prefix}/hello`,
      `${this._prefix}/status`,
      `${this._prefix}/ds/+/evt`,
    ], { qos: 1 })

    // Handle incoming messages
    this._client.on('message', (topic: string, payload: Buffer) => {
      if (this._handler) {
        this._handler(new Uint8Array(payload))
      }
    })

    this._client.on('close', () => {
      this._connected = false
    })

    this._client.on('error', () => {
      this._connected = false
    })
  }

  async disconnect(): Promise<void> {
    if (this._client) {
      await this._client.endAsync()
      this._client = null
    }
    this._connected = false
  }

  async send(packet: Uint8Array): Promise<void> {
    if (!this._client || !this._connected) throw new Error('MQTT: not connected')

    // Extract type byte from wire-encoded packet to determine topic
    if (packet.length >= 4) {
      const type = packet[3]
      const typeHex = type.toString(16).padStart(2, '0')
      const topic = `${this._prefix}/cmd/${typeHex}`
      const qos = this._options.qos ?? 1

      await this._client.publishAsync(topic, Buffer.from(packet), { qos })
    }
  }

  onReceive(handler: (packet: Uint8Array) => void): void {
    this._handler = handler
  }
}

File: sdk/python/src/graft/transports/mqtt.py
"""GRAFT MQTT Transport — aiomqtt"""

from __future__ import annotations

import asyncio
from typing import Callable

try:
    import aiomqtt
except ImportError:
    aiomqtt = None  # type: ignore[assignment]


class MQTTTransport:
    """Async MQTT transport using aiomqtt."""

    def __init__(
        self,
        broker: str,
        port: int = 1883,
        device_id: str = "graft-device",
        username: str | None = None,
        password: str | None = None,
    ) -> None:
        if aiomqtt is None:
            raise ImportError("aiomqtt is required: pip install graft-py[mqtt]")
        self._broker = broker
        self._port = port
        self._device_id = device_id
        self._username = username
        self._password = password
        self._client: object | None = None
        self._connected = False
        self._handler: Callable[[bytes], None] | None = None
        self._task: asyncio.Task[None] | None = None
        self.needs_cobs = False

    @property
    def connected(self) -> bool:
        return self._connected

    async def connect(self) -> None:
        self._client = aiomqtt.Client(
            self._broker,
            port=self._port,
            username=self._username,
            password=self._password,
        )
        await self._client.__aenter__()  # type: ignore[union-attr]
        await self._client.subscribe(f"graft/{self._device_id}/evt/#")  # type: ignore[union-attr]
        self._connected = True
        self._task = asyncio.create_task(self._listen())

    async def disconnect(self) -> None:
        self._connected = False
        if self._client:
            await self._client.__aexit__(None, None, None)  # type: ignore[union-attr]
        self._client = None

    async def send(self, packet: bytes) -> None:
        if not self._connected or not self._client:
            raise RuntimeError("MQTTTransport: not connected")
        await self._client.publish(  # type: ignore[union-attr]
            f"graft/{self._device_id}/cmd",
            payload=packet,
            qos=1,
        )

    def on_receive(self, handler: Callable[[bytes], None]) -> None:
        self._handler = handler

    async def _listen(self) -> None:
        try:
            async for message in self._client.messages:  # type: ignore[union-attr]
                if self._handler and message.payload:
                    self._handler(bytes(message.payload))
        except Exception:
            self._connected = False
